x-pack/plugins/lens/public/datatable_visualization/components/table_basic.tsx

```diff
@@ -120,11 +143,15 @@ export const DatatableComponent = (props: DatatableRenderProps) => {
     (bucketColumns.length &&
       firstTable.rows.every((row) => bucketColumns.every((col) => row[col] == null)));
 
-  const visibleColumns = useMemo(() => columnConfig.columnIds.filter((field) => !!field), [
-    columnConfig,
-  ]);
+  const visibleColumns = useMemo(
+    () =>
+      columnConfig.columns
+        .filter((col) => !!col.columnId && !col.hidden)
+        .map((col) => col.columnId),
+    [columnConfig]
```

Comment:
Maybe there's a JS deep equality issue that causes visibleColumns to get stale when triggered from the visualization?